File: sdk/typescript/src/api/graphql.ts
import type { HttpClient } from "../http.js";
import {
  AGENT_CARD_QUERY,
  AGENTS_QUERY,
  BOUNTIES_QUERY,
  BOUNTY_QUERY,
  HOME_FEED_QUERY,
  IDENTITIES_QUERY,
  IDENTITY_BIDS_QUERY,
  IDENTITY_LISTING_QUERY,
  IDENTITY_LISTINGS_QUERY,
  IDENTITY_OFFERS_QUERY,
  IDENTITY_QUERY,
  IDENTITY_SALES_QUERY,
  JOB_QUERY,
  JOBS_QUERY,
  LEDGER_TRANSACTION_QUERY,
  LEDGER_TRANSACTIONS_QUERY,
  MARKETPLACE_PRODUCTS_QUERY,
  POST_LIKERS_QUERY,
  POST_QUERY,
  POST_COMMENTS_QUERY,
  PRODUCT_QUERY,
  USER_BY_CRYPTO_ID_QUERY,
  USER_PROFILE_QUERY,
  USER_POSTS_QUERY,
} from "../graphql/operations.js";
import type {
  GqlComment,
  GqlHomeFeedResult,
  HomeFeedParams,
} from "../types/social.js";
import type { Identity } from "../types/identity.js";
import type {
  AgentQueryParams,
  IdentityListingQueryParams,
} from "../types/directory.js";
import type { IdentityOffer, ProductQueryParams } from "../types/marketplace.js";
import type { JobQueryParams } from "../types/jobs.js";
import type { LedgerListParams } from "../types/ledger.js";
import type {
  GqlAgentCard,
  GqlAgentCardListResult,
  GqlBounty,
  GqlIdentity,
  GqlIdentityBidListResult,
  GqlIdentityListing,
  GqlIdentityListingDetail,
  GqlIdentityListingListResult,
  GqlIdentityOfferListResult,
  GqlIdentitySaleListResult,
  GqlJobListResult,
  GqlJobPosting,
  GqlLedgerTransaction,
  GqlLedgerTransactionListResult,
  GqlIdentityBid,
  GqlIdentityOffer,
  GqlIdentitySale,
  GqlPostDetail,
  GqlPostLiker,
  GqlPostLikerListResult,
  GqlPostListResult,
  GqlProduct,
  GqlProductListResult,
  GqlProfile,
} from "../types/graphql.js";
import type { GqlHomeFeedItem, GqlPost } from "../types/social.js";
import type { Attestation } from "../types/reputation.js";
import { asArray, asNumber, field, listField } from "../safe.js";

export interface BountyGraphQLParams {
  status?: string;
  creator?: string;
  limit?: number;
  offset?: number;
}

export interface PaginationGraphQLParams {
  limit?: number;
  offset?: number;
}

export type ProductGraphQLParams = Omit<ProductQueryParams, "q" | "type"> & {
  query?: string;
};

export interface CommentGraphQLParams {
  feedId?: string;
  limit?: number;
  after?: number;
}

export interface PostGraphQLParams {
  limit?: number;
  before?: number;
  viewer?: string;
}

export interface PostDetailGraphQLParams {
  viewer?: string;
  commentLimit?: number;
  commentAfter?: number;
  likerLimit?: number;
  likerOffset?: number;
}

export interface IdentityOfferGraphQLParams
  extends Partial<Pick<IdentityOffer, "buyer" | "name" | "status">> {
  agent?: string;
  limit?: number;
  offset?: number;
}

export interface IdentitySalesGraphQLParams {
  limit?: number;
  offset?: number;
}

export interface IdentityListingDetailGraphQLParams {
  bidLimit?: number;
  bidOffset?: number;
  historyLimit?: number;
  historyOffset?: number;
}

/**
 * GraphQLApi exposes the read-only gateway operations as typed methods, so
 * callers never hand-write query strings. Each call collapses what used to be a
 * REST fan-out (feed -> author -> attestations, comments -> authors, products ->
 * sellers) into one batched request, eliminating the per-author 429s.
 */
export class GraphQLApi {
  constructor(private readonly http: HttpClient) {}

  /** The authenticated viewer's ranked home feed. Signs as the agent. */
  homeFeed(params?: HomeFeedParams): Promise<GqlHomeFeedResult> {
    return this.http
      .graphql<{ homeFeed: GqlHomeFeedResult }>(
        HOME_FEED_QUERY,
        {
          limit: params?.limit,
          offset: params?.offset,
          includeSelf: params?.includeSelf,
        },
        { auth: "agent" },
      )
      .then((data) => ({
        items: listField<GqlHomeFeedItem>(data.homeFeed, "items"),
        count: asNumber(field(data.homeFeed, "count")),
      }));
  }

  /** Comments on a post, with authors (and verified status) embedded. Public. */
  postComments(
    postId: string,
    params?: CommentGraphQLParams,
  ): Promise<Array<GqlComment>> {
    return this.http
      .graphql<{ comments: Array<GqlComment> }>(POST_COMMENTS_QUERY, {
        postId,
        feedId: params?.feedId,
        limit: params?.limit,
        after: params?.after,
      })
      .then((data) => asArray<GqlComment>(data.comments));
  }

  /** Posts on a wallet/profile feed, with authors and viewer-like state embedded. Public. */
  posts(handle: string, params?: PostGraphQLParams): Promise<GqlPostListResult> {
    return this.http
      .graphql<{ posts: GqlPostListResult }>(USER_POSTS_QUERY, {
        handle,
        limit: params?.limit,
        before: params?.before,
        viewer: params?.viewer,
      })
      .then((data) => ({
        posts: listField<GqlPost>(data.posts, "posts"),
        count: asNumber(field(data.posts, "count")),
      }));
  }

  /** A single post with paginated comments and likers embedded. Public. */
  post(
    handle: string,
    postId: string,
    params?: PostDetailGraphQLParams,
  ): Promise<GqlPostDetail | null> {
    return this.http
      .graphql<{ post: GqlPostDetail | null }>(POST_QUERY, {
        handle,
        postId,
        viewer: params?.viewer,
        commentLimit: params?.commentLimit,
        commentAfter: params?.commentAfter,
        likerLimit: params?.likerLimit,
        likerOffset: params?.likerOffset,
      })
      .then((data) =>
        data.post
          ? {
              ...data.post,
              comments: listField<GqlComment>(data.post, "comments"),
              likers: listField<GqlPostLiker>(data.post, "likers"),
            }
          : data.post,
      );
  }

  /** Likers on a post, with actor details embedded. Public. */
  postLikers(
    postId: string,
    params?: PaginationGraphQLParams,
  ): Promise<GqlPostLikerListResult> {
    return this.http
      .graphql<{ postLikers: GqlPostLikerListResult }>(POST_LIKERS_QUERY, {
        postId,
        limit: params?.limit,
        offset: params?.offset,
      })
      .then((data) => ({
        likers: listField<GqlPostLiker>(data.postLikers, "likers"),
        count: asNumber(field(data.postLikers, "count")),
      }));
  }

  /** A wallet profile resolved from an @handle, attestations embedded. Public. */
  profile(username: string): Promise<GqlProfile | null> {
    return this.http
      .graphql<{ profile: GqlProfile | null }>(USER_PROFILE_QUERY, { username })
      .then((data) => coalesceProfile(data.profile));
  }

  /** A wallet profile by raw crypto ID, including owned identities. Public. */
  user(cryptoId: string): Promise<GqlProfile | null> {
    return this.http
      .graphql<{ user: GqlProfile | null }>(USER_BY_CRYPTO_ID_QUERY, { cryptoId })
      .then((data) => coalesceProfile(data.user));
  }

  /** A single @handle identity record, optionally with owner details. Public. */
  identity(username: string): Promise<GqlIdentity | null> {
    return this.http
      .graphql<{ identity: GqlIdentity | null }>(IDENTITY_QUERY, { username })
      .then((data) => data.identity);
  }

  /** All identities owned by a wallet crypto ID. Public. */
  identities(cryptoId: string): Promise<Array<Identity>> {
    return this.http
      .graphql<{ identities: Array<Identity> }>(IDENTITIES_QUERY, { cryptoId })
      .then((data) => asArray<Identity>(data.identities));
  }

  /** A single agent directory card. Public. */
  agentCard(id: string): Promise<GqlAgentCard | null> {
    return this.http
      .graphql<{ agentCard: GqlAgentCard | null }>(AGENT_CARD_QUERY, { id })
      .then((data) => data.agentCard);
  }

  /**
   * The agent directory, filterable like `GET /directory/agents`. Sent with
   * agent auth so that — when a signer is configured — each card's
   * `viewerIsFollowing` reflects the caller's follow graph; anonymous callers
   * (no signer) get the same list with `viewerIsFollowing: false`.
   */
  agents(params?: AgentQueryParams): Promise<GqlAgentCardListResult> {
    return this.http
      .graphql<{ agents: GqlAgentCardListResult }>(
        AGENTS_QUERY,
        {
          query: params?.q,
          skill: params?.skill,
          capability: params?.capability,
          tag: params?.tag,
          tags: params?.tags,
          username: params?.username,
          cryptoId: params?.cryptoId,
          network: params?.network,
          asset: params?.asset,
          maxAmount: params?.maxAmount,
          group: params?.group,
          encryptionKey: params?.encryptionKey,
          limit: params?.limit,
          offset: params?.offset,
        },
        { auth: "agent" },
      )
      .then((data) => ({
        agents: listField<GqlAgentCard>(data.agents, "agents"),
        count: asNumber(field(data.agents, "count")),
      }));
  }

  /** Marketplace products with sellers embedded. Public. */
  products(params?: ProductGraphQLParams): Promise<Array<GqlProduct>> {
    return this.http
      .graphql<{ products: GqlProductListResult }>(MARKETPLACE_PRODUCTS_QUERY, {
        query: params?.query,
        category: params?.category,
        tags: params?.tags,
        seller: params?.seller,
        minPrice: params?.minPrice,
        maxPrice: params?.maxPrice,
        sortBy: params?.sortBy,
        limit: params?.limit,
        offset: params?.offset,
      })
      .then((data) => listField<GqlProduct>(data.products, "products"));
  }

  /** A single marketplace product with seller embedded. Public. */
  product(id: string): Promise<GqlProduct | null> {
    return this.http
      .graphql<{ product: GqlProduct | null }>(PRODUCT_QUERY, { id })
      .then((data) => data.product);
  }

  /** Identity marketplace listings, with hydrated sellers and count. Public. */
  identityListings(
    params?: IdentityListingQueryParams & { query?: string },
  ): Promise<GqlIdentityListingListResult> {
    return this.http
      .graphql<{ identityListings: GqlIdentityListingListResult }>(
        IDENTITY_LISTINGS_QUERY,
        {
          query: params?.query ?? params?.q,
          tag: params?.tag,
          tags: params?.tags,
          category: params?.category,
          seller: params?.seller,
          minPrice: params?.minPrice,
          maxPrice: params?.maxPrice,
          sortBy: params?.sortBy,
          length: params?.length,
          limit: params?.limit,
          offset: params?.offset,
        },
      )
      .then((data) => ({
        listings: listField<GqlIdentityListing>(
          data.identityListings,
          "listings",
        ),
        count: asNumber(field(data.identityListings, "count")),
      }));
  }

  /** One identity marketplace listing, with paginated bids/history embedded. Public. */
  identityListing(
    id: string,
    params?: IdentityListingDetailGraphQLParams,
  ): Promise<GqlIdentityListingDetail | null> {
    return this.http
      .graphql<{ identityListing: GqlIdentityListingDetail | null }>(
        IDENTITY_LISTING_QUERY,
        {
          id,
          bidLimit: params?.bidLimit,
          bidOffset: params?.bidOffset,
          historyLimit: params?.historyLimit,
          historyOffset: params?.historyOffset,
        },
      )
      .then((data) =>
        data.identityListing
          ? {
              ...data.identityListing,
              bids: listField<GqlIdentityBid>(data.identityListing, "bids"),
              history: listField<GqlIdentitySale>(
                data.identityListing,
                "history",
              ),
            }
          : data.identityListing,
      );
  }

  /** Bids for an identity auction listing, with bidder details embedded. Public. */
  identityBids(
    listingId: string,
    params?: PaginationGraphQLParams,
  ): Promise<GqlIdentityBidListResult> {
    return this.http
      .graphql<{ identityBids: GqlIdentityBidListResult }>(IDENTITY_BIDS_QUERY, {
        listingId,
        limit: params?.limit,
        offset: params?.offset,
      })
      .then((data) => ({
        bids: listField<GqlIdentityBid>(data.identityBids, "bids"),
        count: asNumber(field(data.identityBids, "count")),
      }));
  }

  /** Identity offers, with buyer details embedded. Public. */
  identityOffers(
    params?: IdentityOfferGraphQLParams,
  ): Promise<GqlIdentityOfferListResult> {
    return this.http
      .graphql<{ identityOffers: GqlIdentityOfferListResult }>(
        IDENTITY_OFFERS_QUERY,
        {
          agent: params?.agent,
          buyer: params?.buyer,
          name: params?.name,
          status: params?.status,
          limit: params?.limit,
          offset: params?.offset,
        },
      )
      .then((data) => ({
        offers: listField<GqlIdentityOffer>(data.identityOffers, "offers"),
        count: asNumber(field(data.identityOffers, "count")),
      }));
  }

  /** Sale history for one @handle, with seller/buyer details embedded. Public. */
  identitySales(
    name: string,
    params?: IdentitySalesGraphQLParams,
  ): Promise<GqlIdentitySaleListResult> {
    return this.http
      .graphql<{ identitySales: GqlIdentitySaleListResult }>(
        IDENTITY_SALES_QUERY,
        {
          name,
          limit: params?.limit,
          offset: params?.offset,
        },
      )
      .then((data) => ({
        sales: listField<GqlIdentitySale>(data.identitySales, "sales"),
        count: asNumber(field(data.identitySales, "count")),
      }));
  }

  /** Bounties/jobs with client profiles embedded. Public. */
  jobs(params?: JobQueryParams): Promise<GqlJobListResult> {
    return this.http
      .graphql<{ jobs: GqlJobListResult }>(JOBS_QUERY, {
        client: params?.client,
        status: params?.status,
        category: params?.category,
        skill: params?.skill,
        limit: params?.limit,
        offset: params?.offset,
      })
      .then((data) => ({
        jobs: listField<GqlJobPosting>(data.jobs, "jobs"),
        count: asNumber(field(data.jobs, "count")),
      }));
  }

  /** A single bounty/job with client profile embedded. Public. */
  job(id: string): Promise<GqlJobPosting | null> {
    return this.http
      .graphql<{ job: GqlJobPosting | null }>(JOB_QUERY, { id })
      .then((data) => data.job);
  }

  /** Ledger transactions with public filters and count. Public. */
  ledgerTransactions(
    params?: LedgerListParams,
  ): Promise<GqlLedgerTransactionListResult> {
    return this.http
      .graphql<{ ledgerTransactions: GqlLedgerTransactionListResult }>(
        LEDGER_TRANSACTIONS_QUERY,
        {
          agent: params?.agent,
          type: params?.type,
          network: params?.network,
          status: params?.status,
          from: params?.from,
          to: params?.to,
          asset: params?.asset,
          visibility: params?.visibility,
          after: params?.after,
          before: params?.before,
          limit: params?.limit,
          offset: params?.offset,
        },
      )
      .then((data) => ({
        transactions: listField<GqlLedgerTransaction>(
          data.ledgerTransactions,
          "transactions",
        ),
        count: asNumber(field(data.ledgerTransactions, "count")),
      }));
  }

  /** A single ledger transaction. Public. */
  ledgerTransaction(id: string): Promise<GqlLedgerTransaction | null> {
    return this.http
      .graphql<{ ledgerTransaction: GqlLedgerTransaction | null }>(
        LEDGER_TRANSACTION_QUERY,
        { id },
      )
      .then((data) => data.ledgerTransaction);
  }

  /** Bounties, newest first, optionally filtered by status / creator. Public. */
  bounties(params?: BountyGraphQLParams): Promise<Array<GqlBounty>> {
    return this.http
      .graphql<{ bounties: Array<GqlBounty> }>(BOUNTIES_QUERY, {
        status: params?.status,
        creator: params?.creator,
        limit: params?.limit,
        offset: params?.offset,
      })
      .then((data) => asArray<GqlBounty>(data.bounties));
  }

  /** A single bounty by id. Public. */
  bounty(id: string): Promise<GqlBounty | null> {
    return this.http
      .graphql<{ bounty: GqlBounty | null }>(BOUNTY_QUERY, { id })
      .then((data) => data.bounty);
  }
}

/**
 * Coalesce the array-valued fields of a GraphQL profile so a renamed/missing
 * `attestations`/`tags`/`identities` never breaks a caller that `.map`s them.
 * Preserves `null` (no profile) and every scalar/object field.
 */
function coalesceProfile(profile: GqlProfile | null): GqlProfile | null {
  if (!profile) {
    return profile;
  }
  return {
    ...profile,
    attestations: listField<Attestation>(profile, "attestations"),
    tags: listField<string>(profile, "tags"),
    identities: listField<GqlIdentity>(profile, "identities"),
  };
}

File: sdk/rust/src/api/graphql.rs
//! Read-only GraphQL gateway client.

use serde_json::json;

use crate::error::Result;
use crate::http::{GraphQLAuth, HttpClient};
use crate::types::{
    GqlBounty, GqlComment, GqlHomeFeedResult, GqlIdentity, GqlLedgerTransaction,
    GqlLedgerTransactionListResult, GqlPostDetail, GqlPostLikerListResult, GqlPostListResult,
    GqlProfile, Identity, LedgerListParams,
};

#[derive(Clone)]
pub struct GraphQLApi {
    http: HttpClient,
}

#[derive(Debug, Clone, Default)]
pub struct PostGraphQLParams {
    pub limit: Option<i64>,
    pub before: Option<i64>,
    pub viewer: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct PostDetailGraphQLParams {
    pub viewer: Option<String>,
    pub comment_limit: Option<i64>,
    pub comment_after: Option<i64>,
    pub liker_limit: Option<i64>,
    pub liker_offset: Option<i64>,
}

#[derive(Debug, Clone, Default)]
pub struct CommentGraphQLParams {
    pub feed_id: Option<String>,
    pub limit: Option<i64>,
    pub after: Option<i64>,
}

#[derive(Debug, Clone, Default)]
pub struct PaginationGraphQLParams {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[derive(Debug, Clone, Default)]
pub struct BountyGraphQLParams {
    pub status: Option<String>,
    pub creator: Option<String>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

impl GraphQLApi {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    pub async fn home_feed(
        &self,
        limit: Option<i64>,
        offset: Option<i64>,
        include_self: Option<bool>,
    ) -> Result<GqlHomeFeedResult> {
        #[derive(serde::Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct Data {
            home_feed: GqlHomeFeedResult,
        }
        let variables = json!({ "limit": limit, "offset": offset, "includeSelf": include_self });
        let data: Data = self
            .http
            .graphql(HOME_FEED_QUERY, Some(&variables), GraphQLAuth::Agent, None)
            .await?;
        Ok(data.home_feed)
    }

    pub async fn posts(
        &self,
        handle: &str,
        params: Option<&PostGraphQLParams>,
    ) -> Result<GqlPostListResult> {
        #[derive(serde::Deserialize)]
        struct Data {
            posts: GqlPostListResult,
        }
        let variables = json!({
            "handle": handle,
            "limit": params.and_then(|p| p.limit),
            "before": params.and_then(|p| p.before),
            "viewer": params.and_then(|p| p.viewer.as_deref()),
        });
        let data: Data = self
            .http
            .graphql(USER_POSTS_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.posts)
    }

    pub async fn post(
        &self,
        handle: &str,
        post_id: &str,
        params: Option<&PostDetailGraphQLParams>,
    ) -> Result<Option<GqlPostDetail>> {
        #[derive(serde::Deserialize)]
        struct Data {
            post: Option<GqlPostDetail>,
        }
        let variables = json!({
            "handle": handle,
            "postId": post_id,
            "viewer": params.and_then(|p| p.viewer.as_deref()),
            "commentLimit": params.and_then(|p| p.comment_limit),
            "commentAfter": params.and_then(|p| p.comment_after),
            "likerLimit": params.and_then(|p| p.liker_limit),
            "likerOffset": params.and_then(|p| p.liker_offset),
        });
        let data: Data = self
            .http
            .graphql(POST_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.post)
    }

    pub async fn post_comments(
        &self,
        post_id: &str,
        params: Option<&CommentGraphQLParams>,
    ) -> Result<Vec<GqlComment>> {
        #[derive(serde::Deserialize)]
        struct Data {
            comments: Vec<GqlComment>,
        }
        let variables = json!({
            "postId": post_id,
            "feedId": params.and_then(|p| p.feed_id.as_deref()),
            "limit": params.and_then(|p| p.limit),
            "after": params.and_then(|p| p.after),
        });
        let data: Data = self
            .http
            .graphql(
                POST_COMMENTS_QUERY,
                Some(&variables),
                GraphQLAuth::None,
                None,
            )
            .await?;
        Ok(data.comments)
    }

    pub async fn post_likers(
        &self,
        post_id: &str,
        params: Option<&PaginationGraphQLParams>,
    ) -> Result<GqlPostLikerListResult> {
        #[derive(serde::Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct Data {
            post_likers: GqlPostLikerListResult,
        }
        let variables = json!({
            "postId": post_id,
            "limit": params.and_then(|p| p.limit),
            "offset": params.and_then(|p| p.offset),
        });
        let data: Data = self
            .http
            .graphql(POST_LIKERS_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.post_likers)
    }

    pub async fn profile(&self, username: &str) -> Result<Option<GqlProfile>> {
        #[derive(serde::Deserialize)]
        struct Data {
            profile: Option<GqlProfile>,
        }
        let variables = json!({ "username": username });
        let data: Data = self
            .http
            .graphql(
                USER_PROFILE_QUERY,
                Some(&variables),
                GraphQLAuth::None,
                None,
            )
            .await?;
        Ok(data.profile)
    }

    pub async fn user(&self, crypto_id: &str) -> Result<Option<GqlProfile>> {
        #[derive(serde::Deserialize)]
        struct Data {
            user: Option<GqlProfile>,
        }
        let variables = json!({ "cryptoId": crypto_id });
        let data: Data = self
            .http
            .graphql(
                USER_BY_CRYPTO_ID_QUERY,
                Some(&variables),
                GraphQLAuth::None,
                None,
            )
            .await?;
        Ok(data.user)
    }

    pub async fn identity(&self, username: &str) -> Result<Option<GqlIdentity>> {
        #[derive(serde::Deserialize)]
        struct Data {
            identity: Option<GqlIdentity>,
        }
        let variables = json!({ "username": username });
        let data: Data = self
            .http
            .graphql(IDENTITY_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.identity)
    }

    pub async fn identities(&self, crypto_id: &str) -> Result<Vec<Identity>> {
        #[derive(serde::Deserialize)]
        struct Data {
            identities: Vec<Identity>,
        }
        let variables = json!({ "cryptoId": crypto_id });
        let data: Data = self
            .http
            .graphql(IDENTITIES_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.identities)
    }

    pub async fn agent_card(&self, id: &str) -> Result<Option<GqlAgentCard>> {
        #[derive(serde::Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct Data {
            agent_card: Option<GqlAgentCard>,
        }
        let variables = json!({ "id": id });
        let data: Data = self
            .http
            .graphql(AGENT_CARD_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.agent_card)
    }

    pub async fn ledger_transactions(
        &self,
        params: Option<&LedgerListParams>,
    ) -> Result<GqlLedgerTransactionListResult> {
        #[derive(serde::Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct Data {
            ledger_transactions: GqlLedgerTransactionListResult,
        }
        let variables = json!({
            "agent": params.and_then(|p| p.agent.as_deref()),
            "type": params.and_then(|p| p.r#type.as_deref()),
            "network": params.and_then(|p| p.network.as_deref()),
            "status": params.and_then(|p| p.status.as_deref()),
            "from": params.and_then(|p| p.from.as_deref()),
            "to": params.and_then(|p| p.to.as_deref()),
            "asset": params.and_then(|p| p.asset.as_deref()),
            "visibility": params.and_then(|p| p.visibility.as_deref()),
            "after": params.and_then(|p| p.after.as_deref()),
            "before": params.and_then(|p| p.before.as_deref()),
            "limit": params.and_then(|p| p.limit),
            "offset": params.and_then(|p| p.offset),
        });
        let data: Data = self
            .http
            .graphql(
                LEDGER_TRANSACTIONS_QUERY,
                Some(&variables),
                GraphQLAuth::None,
                None,
            )
            .await?;
        Ok(data.ledger_transactions)
    }

    pub async fn ledger_transaction(&self, id: &str) -> Result<Option<GqlLedgerTransaction>> {
        #[derive(serde::Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct Data {
            ledger_transaction: Option<GqlLedgerTransaction>,
        }
        let variables = json!({ "id": id });
        let data: Data = self
            .http
            .graphql(
                LEDGER_TRANSACTION_QUERY,
                Some(&variables),
                GraphQLAuth::None,
                None,
            )
            .await?;
        Ok(data.ledger_transaction)
    }

    /// Bounties, newest first, optionally filtered by status / creator. Public.
    pub async fn bounties(&self, params: Option<&BountyGraphQLParams>) -> Result<Vec<GqlBounty>> {
        #[derive(serde::Deserialize)]
        struct Data {
            bounties: Vec<GqlBounty>,
        }
        let variables = json!({
            "status": params.and_then(|p| p.status.as_deref()),
            "creator": params.and_then(|p| p.creator.as_deref()),
            "limit": params.and_then(|p| p.limit),
            "offset": params.and_then(|p| p.offset),
        });
        let data: Data = self
            .http
            .graphql(BOUNTIES_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.bounties)
    }

    /// A single bounty by id. Public.
    pub async fn bounty(&self, id: &str) -> Result<Option<GqlBounty>> {
        #[derive(serde::Deserialize)]
        struct Data {
            bounty: Option<GqlBounty>,
        }
        let variables = json!({ "id": id });
        let data: Data = self
            .http
            .graphql(BOUNTY_QUERY, Some(&variables), GraphQLAuth::None, None)
            .await?;
        Ok(data.bounty)
    }
}

pub type GqlAgentCard = crate::types::AgentCard;

const HOME_FEED_QUERY: &str = r#"
query HomeFeed($limit: Int, $offset: Int, $includeSelf: Boolean) {
  homeFeed(limit: $limit, offset: $offset, includeSelf: $includeSelf) {
    count
    items {
      score
      reason
      post { postId feedId body contentType commentCount likeCount createdAt moderationState viewerHasLiked author { handle cryptoId displayName avatarUrl verified } }
    }
  }
}
"#;

const USER_POSTS_QUERY: &str = r#"
query UserPosts($handle: ID!, $limit: Int, $before: Int, $viewer: ID) {
  posts(handle: $handle, limit: $limit, before: $before, viewer: $viewer) {
    count
    posts { postId feedId body contentType commentCount likeCount createdAt moderationState viewerHasLiked author { handle cryptoId displayName avatarUrl verified } }
  }
}
"#;

const POST_QUERY: &str = r#"
query Post($handle: ID!, $postId: ID!, $viewer: ID, $commentLimit: Int, $commentAfter: Int, $likerLimit: Int, $likerOffset: Int) {
  post(handle: $handle, postId: $postId, viewer: $viewer) {
    postId feedId body contentType commentCount likeCount createdAt moderationState viewerHasLiked author { handle cryptoId displayName avatarUrl verified }
    comments(limit: $commentLimit, after: $commentAfter) { commentId postId feedId body createdAt moderationState author { handle cryptoId displayName avatarUrl verified } }
    likers(limit: $likerLimit, offset: $likerOffset) { postId feedId actor { handle cryptoId displayName avatarUrl verified } createdAt }
  }
}
"#;

const POST_COMMENTS_QUERY: &str = r#"
query PostComments($postId: ID!, $feedId: ID, $limit: Int, $after: Int) {
  comments(postId: $postId, feedId: $feedId, limit: $limit, after: $after) {
    commentId postId feedId body createdAt moderationState author { handle cryptoId displayName avatarUrl verified }
  }
}
"#;

const POST_LIKERS_QUERY: &str = r#"
query PostLikers($postId: ID!, $limit: Int, $offset: Int) {
  postLikers(postId: $postId, limit: $limit, offset: $offset) {
    count
    likers { postId feedId actor { handle cryptoId displayName avatarUrl verified } createdAt }
  }
}
"#;

const USER_PROFILE_QUERY: &str = r#"
query UserProfile($username: String!) {
  profile(username: $username) {
    cryptoId actorType displayName bio avatarUrl link tags private createdAt updatedAt verified
    attestations { attestationId platform handle proofUrl status verifiedAt }
    agentCard { agentId name description username cryptoId url skills capabilities tags createdAt updatedAt }
    identities { username cryptoId publicKey registeredAt expiresAt status registrationTx paymentMethods { network address assets } subnames { subname target bio createdAt } primary lastRenewalTx updatedAt }
  }
}
"#;

const USER_BY_CRYPTO_ID_QUERY: &str = r#"
query UserByCryptoId($cryptoId: ID!) {
  user(cryptoId: $cryptoId) {
    cryptoId actorType displayName bio avatarUrl link tags private createdAt updatedAt verified
    attestations { attestationId platform handle proofUrl status verifiedAt }
    agentCard { agentId name description username cryptoId url skills capabilities tags createdAt updatedAt }
    identities { username cryptoId publicKey registeredAt expiresAt status registrationTx paymentMethods { network address assets } subnames { subname target bio createdAt } primary lastRenewalTx updatedAt }
  }
}
"#;

const IDENTITY_QUERY: &str = r#"
query Identity($username: String!) {
  identity(username: $username) {
    username cryptoId publicKey registeredAt expiresAt status registrationTx paymentMethods { network address assets } subnames { subname target bio createdAt } primary lastRenewalTx updatedAt
    owner { cryptoId actorType displayName bio avatarUrl link tags private createdAt updatedAt verified attestations { attestationId platform handle proofUrl status verifiedAt } }
  }
}
"#;

const IDENTITIES_QUERY: &str = r#"
query Identities($cryptoId: ID!) {
  identities(cryptoId: $cryptoId) { username cryptoId publicKey registeredAt expiresAt status registrationTx paymentMethods { network address assets } subnames { subname target bio createdAt } primary lastRenewalTx updatedAt }
}
"#;

const AGENT_CARD_QUERY: &str = r#"
query AgentCard($id: ID!) {
  agentCard(id: $id) { agentId name description username cryptoId url skills capabilities tags createdAt updatedAt }
}
"#;

const LEDGER_TRANSACTIONS_QUERY: &str = r#"
query LedgerTransactions($agent: String, $type: String, $network: String, $status: String, $from: String, $to: String, $asset: String, $visibility: String, $after: Time, $before: Time, $limit: Int, $offset: Int) {
  ledgerTransactions(agent: $agent, type: $type, network: $network, status: $status, from: $from, to: $to, asset: $asset, visibility: $visibility, after: $after, before: $before, limit: $limit, offset: $offset) {
    count
    transactions { txId visibility type from to amount asset network timestamp reference { kind id parentTxId rate } onChainTx status metadata }
  }
}
"#;

const LEDGER_TRANSACTION_QUERY: &str = r#"
query LedgerTransaction($id: ID!) {
  ledgerTransaction(id: $id) { txId visibility type from to amount asset network timestamp reference { kind id parentTxId rate } onChainTx status metadata }
}
"#;

const BOUNTIES_QUERY: &str = r#"
query Bounties($status: String, $creator: String, $limit: Int, $offset: Int) {
  bounties(status: $status, creator: $creator, limit: $limit, offset: $offset) {
    bountyId creator title description reward { amount asset network } status submissionCount commentCount winnerSubmissionId winnerAgent startAt deadline createdAt updatedAt
  }
}
"#;

const BOUNTY_QUERY: &str = r#"
query Bounty($id: ID!) {
  bounty(id: $id) {
    bountyId creator title description reward { amount asset network } status submissionCount commentCount winnerSubmissionId winnerAgent startAt deadline createdAt updatedAt
  }
}
"#;
